Ilariae/solana shims (#532)
wormhole-foundation/wormhole-docs: products/messaging/concepts/solana-shim.md

---
title: Solana Shims
description: Understand how Wormhole uses shim programs on Solana to optimize message emission and VAA verification without modifying the Core Bridge.
categories: Basics
---

# Solana Shims

Wormhole shims on Solana are lightweight programs that enable cheaper and more flexible message emission and verification while preserving Guardian observation guarantees. They are designed for integrators who want to reduce Solana rent costs without sacrificing core protocol security or Guardian compatibility.

## The Core Bridge Account Problem

> When you emit a message on Solana using the legacy [Wormhole core bridge](/docs/protocol/infrastructure/core-contracts/){target=\_blank}, it creates a new on-chain account, a Program Derived Address (PDA), for every message. Each of these accounts must hold enough SOL to be rent-exempt, locking up lamports that cannot be reclaimed since core bridge does not allow these accounts to be closed. Over time, this results in two big problems:

Reviewer: ```suggestion
When you emit a message on Solana using the legacy [Wormhole core bridge](/docs/protocol/infrastructure/core-contracts/){target=\_blank}, it creates a new on-chain account, a Program Derived Address (PDA), for every message. Each of these accounts must hold enough SOL to be rent-exempt, locking up lamports that cannot be reclaimed since the core bridge does not allow these accounts to be closed. Over time, this results in two big problems:
```

> When you emit a message on Solana using the legacy [Wormhole core bridge](/docs/protocol/infrastructure/core-contracts/){target=\_blank}, it creates a new on-chain account — a Program Derived Address (PDA) — for every message. Each of these accounts must hold enough SOL to be rent-exempt, locking up lamports that cannot be reclaimed since core bridge does not allow these accounts to be closed. Over time, this results in two big problems:

Reviewer 2: ```suggestion
When you emit a message on Solana using the legacy [Wormhole core bridge](/docs/protocol/infrastructure/core-contracts/){target=\_blank}, it creates a new on-chain account, a Program Derived Address (PDA), for every message. Each of these accounts must hold enough SOL to be rent-exempt, locking up lamports that cannot be reclaimed since core bridge does not allow these accounts to be closed. Over time, this results in two big problems:
```

- **Permanent On-Chain State**: Every message leaves behind a permanent account, increasing long-term storage needs on Solana.
- **Lost Lamports to Rent**: Integrators lose SOL for every message, as the lamports needed for rent exemption remain locked in the message accounts indefinitely.

Solana’s rent-exemption model isn't the fundamental limitation; the constraint lies in the legacy `post_message` function of the core bridge, which always creates a new, non-reclaimable account every time it’s called. Even after a message is consumed, these accounts can’t be closed or reused, resulting in unrecoverable rent costs.

> Although the `post_message_unreliable` function allows for account reuse, it comes with strict tradeoffs. Once a message is overwritten, it can no longer be recovered, making it non-re-observable if missed by Guardians. It also locks you into the original account size, as the feature predates Solana’s account resizing.

Reviewer: non-re-observable is a bit weird to say, is there another word or way to say this?

This design ensures reliability, as message data is always available on-chain for Guardians to observe. However, it comes at a cost in both storage and lost SOL. To address these issues, Wormhole introduces Solana shims, which fundamentally change the cost model for emission and verification.

## What Are the Solana Shims?

To address the limitations of the core bridge, Wormhole deploys two specialized Solana programs called shims:

- **[Post Message Shim (`EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX`)](https://explorer.solana.com/address/EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX){target=\_blank}**: Emits Wormhole messages efficiently, without creating new message accounts for each emission, reducing rent costs.
- **[Verify VAA Shim (`EFaNWErqAtVWufdNb7yofSHHfWFos843DFpu4JBw24at`)](https://explorer.solana.com/address/EFaNWErqAtVWufdNb7yofSHHfWFos843DFpu4JBw24at){target=\_blank}**: Verifies VAAs on-chain without leaving permanent accounts.

Both act as lightweight wrappers around the existing core bridge. There are two different options, depending on whether you are emitting messages or verifying VAAs:

### Emission Shim

> The [Emission Shim](/docs/products/messaging/guides/solana-shims/sol-emission/){target=\_blank} is a Solana program deployed at [`EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX`](https://explorer.solana.com/address/EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX){target=\_blank}. It wraps the core bridge’s `post_message_unreliable` instruction and emits message data as a log event instead of storing it in a rent-exempt message account. This removes rent costs and avoids long-term state bloat. Guardians are configured to observe this canonical shim, allowing integrators to send messages through it without additional setup.

Reviewer: ```suggestion
The [Emission Shim](/docs/products/messaging/guides/solana-shims/sol-emission/){target=\_blank} is a Solana program deployed at [`EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX`](https://explorer.solana.com/address/EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX){target=\_blank}. It wraps the core bridge’s `post_message_unreliable` instruction and emits message data as a log event instead of storing it in a rent-exempt message account. This eliminates rent costs and prevents long-term state bloat. Guardians are configured to observe this canonical shim, allowing integrators to send messages through it without additional setup.
```

It works by calling the [`post_message`](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/idls/wormhole_post_message_shim.json){target=_blank} instruction on the Post Message Shim program. This emits the Wormhole message as a log event instead of creating a rent-exempt message account.

> The shim differs from the standard `post_message` approach in two key ways. First, it uses a Program Derived Address (PDA) per emitter for message accounts, removing the need to generate a new keypair for each emission. Second, instead of writing the message into a persistent, rent-exempt account, it emits the data via an Anchor CPI event, which Guardians can observe directly. This design reduces rent costs and avoids leaving unused accounts behind.

Reviewer: ```suggestion
The shim differs from the standard `post_message` approach in two key ways. First, it utilizes a Program Derived Address (PDA) per emitter for message accounts, eliminating the need to generate a new key pair for each emission. Second, instead of writing the message into a persistent, rent-exempt account, it emits the data via an Anchor CPI event, which Guardians can observe directly. This design reduces rent costs and prevents unused accounts from being left behind.
```

> The shim differs from the standard `post_message` approach in two key ways. First, it uses a Program Derived Address (PDA) per emitter for message accounts, removing the need to generate a new keypair for each emission. SSecond, instead of writing the message into a persistent, rent-exempt account, it emits the data via an Anchor CPI event, which Guardians can observe directly. This design reduces rent costs and avoids leaving unused accounts behind.

Reviewer 2: ```suggestion
The shim differs from the standard `post_message` approach in two key ways. First, it uses a Program Derived Address (PDA) per emitter for message accounts, removing the need to generate a new keypair for each emission. Second, instead of writing the message into a persistent, rent-exempt account, it emits the data via an Anchor CPI event, which Guardians can observe directly. This design reduces rent costs and avoids leaving unused accounts behind.
```

The shim works through a few main components:

- **Shim Program**: Provides a `post_message` instruction modeled on the core bridge’s `post_message_unreliable`.

> - **Sequence Handling**: The core bridge still manages sequence numbers. It reads the sequence number from the core bridge and emits it in a [CPI event](https://www.anchor-lang.com/docs/basics/cpi){target=\_blank}, along with the timestamp.

Reviewer: ```suggestion
- **Sequence Handling**: The core bridge continues to manage sequence numbers. It reads the sequence number from the core bridge and emits it in a [CPI event](https://www.anchor-lang.com/docs/basics/cpi){target=\_blank}, along with the timestamp.
```

### Verification Shim 

The [Verification Shim](/docs/products/messaging/guides/solana-shims/sol-verification/){target=\_blank} is a Solana program deployed at [`EFaNWErqAtVWufdNb7yofSHHfWFos843DFpu4JBw24at`](https://explorer.solana.com/address/EFaNWErqAtVWufdNb7yofSHHfWFos843DFpu4JBw24at){target=\_blank}. It provides a [`verify_hash`](https://github.com/wormhole-foundation/wormhole/blob/4656bd4a72cb99f4e94a771a802856c9451af844/svm/wormhole-core-shims/programs/verify-vaa/src/lib.rs#L195){target=\_blank} instruction that checks Guardian signatures against the active Guardian set for a VAA's digest. It ensures quorum, validates each signature in order, recovers the public keys, and matches them against the Guardian set. If all checks pass, the VAA is verified without creating persistent rent-exempt accounts. This replaces using the core bridge’s `post_vaa`. Integrators can call the canonical shim, but existing programs may need to be modified to adopt this approach.

> It works by calling first, the [`post_signatures`](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/idls/wormhole_verify_vaa_shim.json#L43){target=_blank} on the Verification Shim to store Guardian signatures in a temporary account. Then, from within your program, call [`verify_hash`](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/programs/verify-vaa/README.md#verify-hash-technical-details){target=_blank} to check the VAA’s digest against Guardian signatures. In the same transaction, close the signatures account with [`close_signatures`](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/idls/wormhole_verify_vaa_shim.json#L11){target=\_blank} to reclaim rent. 

Reviewer: ```suggestion
It works by first calling the [`post_signatures`](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/idls/wormhole_verify_vaa_shim.json#L43){target=_blank} on the Verification Shim to store Guardian signatures in a temporary account. Then, from within your program, call [`verify_hash`](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/programs/verify-vaa/README.md#verify-hash-technical-details){target=_blank} to check the VAA’s digest against Guardian signatures. In the same transaction, close the signatures account with [`close_signatures`](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/idls/wormhole_verify_vaa_shim.json#L11){target=\_blank} to reclaim rent. 
```

Instead of the core bridge instructions, such as `verify_signatures` and `post_vaa`, the verification shim provides its own flow using `post_signatures`, `verify_hash`, and `close_signatures`. The flow is a simpler sequence that avoids leaving permanent accounts on-chain:

1. Call `post_signatures`: Creates (or appends to) a temporary `GuardianSignatures` account that stores the collected Guardian signatures. This account is owned and managed by the verification shim.
2. Call `verify_hash`: Verifies the digest of the VAA against the active Guardian set and checks quorum by recovering and validating each Guardian signature. If verification succeeds, your program can continue its logic.

> 3. Call `close_signatures`: Immediately close the `GuardianSignatures` account to reclaim the lamports paid for its creation.

Reviewer: ```suggestion
3. Call `close_signatures`: Immediately closes the `GuardianSignatures` account to reclaim the lamports paid for its creation.
```

## Guardian Observation Methods

|                      | Legacy Model           | Shim Model               |
|----------------------|------------------------|--------------------------|
| Message Storage      | On-chain account       | Transaction logs (CPI)   |
| Data Permanence      | Permanent              | Until RPC history pruned |
| Guardian Observation | Reads account data     | Reads transaction logs   |
| Cost                 | High (rent + compute)  | Low (compute only)       |
| Closing Accounts     | Not possible           | Not needed               |

With shims, the message’s existence depends on the transaction log, so cost drops, but indefinite on-chain visibility is no longer guaranteed. Sequence tracking remains the same as the legacy model, so integrators can switch between the two without disrupting sequence numbers.

## Transaction Costs

Solana charges for two main resources when processing transactions: compute units (for execution) and rent (for storing data on-chain). Understanding how each contributes to the overall cost is key to seeing why shims are cheaper.

> - **Compute Units (CU)**: Solana measures CPU resource usage per transaction as “compute units”. Each transaction has a CU limit (usually ~200,000, which can be increased for a fee).

Reviewer: ```suggestion
- **Compute Units (CU)**: Solana measures CPU resource usage per transaction as “compute units”. Each transaction has a CU limit (usually ~200,000 — which can be increased for a fee).
```

- **Rent**: One-time cost in SOL to keep an account on-chain. Most of the core bridge’s cost comes from rent, not CUs.

Even though the shim uses slightly more compute (extra logic for logging), it avoids account creation entirely. Since rent is the most significant cost, the total emission cost drops.

## Safety, Tradeoffs & Limitations

> Shims preserve the same security guarantees as the core bridge, so integrators can adopt shims without weakening protocol safety. The only difference is where data lives: instead of being stored permanently in message accounts, it is emitted in transaction logs or held temporarily until verification completes. Guardians are explicitly configured to observe shim output, ensuring messages and VAAs remain verifiable across the network.

Reviewer: ```suggestion
Shims preserve the same security guarantees as the core bridge so that integrators can adopt shims without weakening protocol safety. The only difference is where data lives: instead of being stored permanently in message accounts, it is emitted in transaction logs or held temporarily until verification completes. Guardians are explicitly configured to observe shim output, ensuring messages and VAAs remain verifiable across the network.
```

> The main tradeoff is durability. In the legacy model, messages and VAAs were always available on-chain for re-observation. With shims, message data persists only as long as transaction history is retained by RPC providers. This is sufficient for Guardian observation but doesn’t provide indefinite public access to raw message data. Applications that rely on long-term on-chain storage may still prefer the legacy path, while most integrators benefit from the reduced cost and state bloat.

Reviewer: ```suggestion
The main tradeoff is durability. In the legacy model, messages and VAAs were always available on-chain for re-observation. With shims, message data persists only as long as transaction history is retained by RPC providers. This is sufficient for Guardian observation, but doesn’t provide indefinite public access to raw message data. Applications that rely on long-term on-chain storage may still prefer the legacy path, while most integrators benefit from the reduced cost and state bloat.
```

---
title: Solana Shims
description: Understand how Wormhole uses shim programs on Solana to optimize message emission and VAA verification without modifying the Core Bridge.
categories: Basics
---

# Solana Shims

Wormhole shims on Solana are lightweight programs that enable cheaper and more flexible message emission and verification while preserving Guardian observation guarantees. They are designed for integrators who want to reduce Solana rent costs without sacrificing core protocol security or Guardian compatibility.

This page explains what shims are, why they were created, how they work, and what this means for integrators.

## The Core Bridge Account Problem

> When you emit a message on Solana using the legacy [Wormhole Core Bridge](/docs/protocol/infrastructure/core-contracts/){target=\_blank}, it creates a new on-chain account—specifically, a PDA (Program Derived Address) account—for every message. These accounts must be rent-exempt, meaning SOL is locked up for each one and cannot be recovered, since Core Bridge does not allow these accounts to be closed. Over time, this results in two big problems:

Reviewer 2: ```suggestion
When you emit a message on Solana using the legacy [Wormhole Core Bridge](/docs/protocol/infrastructure/core-contracts/){target=\_blank}, it creates a new on-chain account, specifically, a PDA (Program Derived Address) account, for every message. These accounts must be rent-exempt, meaning SOL is locked up for each one and cannot be recovered, since Core Bridge does not allow these accounts to be closed. Over time, this results in two big problems:
```

- **Permanent On-Chain State**: Every message leaves behind a permanent account, increasing long-term storage needs on Solana.
- **Lost Lamports to Rent**: Integrators lose SOL for every message, with no way to recover it even after the message has served its purpose. 

> Solana doesn’t use gas like EVM chains. Instead, programs must prepay “rent” in SOL for each account. These lamports can only be reclaimed if the account is closed, which isn’t possible for Core Bridge message accounts.

Reviewer 3: Lets adjust this part, as its not completely correct. 
The statement implies programs must "prepay rent", in reality, accounts need to maintain rent exemption by holding sufficient lamports. 
Also this is not a limitation of Solana, but rather a limitation of the legacy post_message function. 
That function created a new message account for every emission / call. These accounts could not be reclaimed and took up state forever, even long after the VAA has been produced and consumed.

While `post_message_unreliable` allows for account reuse, it has strict limitations: overwritten messages are lost forever, and account reuse can cause sequence conflicts or duplicate VAAs. Guardians expect every message to have a unique emitter and sequence, so reusing accounts risks breaking core protocol guarantees.

Verification adds even more cost: the `post_vaa` instruction creates even more temporary accounts for signatures and VAA data—each adding further rent costs and storage overhead. These accounts aren’t automatically cleaned up, so the cost and on-chain state only grow with usage.

> This design does ensure reliability—messages and verification data are always available on-chain for Guardians to observe. However, it comes at a cost in both storage and lost SOL. To address these issues, Wormhole introduces Solana shims, which fundamentally change the cost model for emission and verification:

Reviewer 2: ```suggestion
This design does ensure reliability, messages and verification data are always available on-chain for Guardians to observe. However, it comes at a cost in both storage and lost SOL. To address these issues, Wormhole introduces Solana shims, which fundamentally change the cost model for emission and verification.
```

## What Are the Solana Shim Contracts?

To address the limitations of the Core Bridge, Wormhole deploys two specialized Solana programs called shims:

> - **Post Message Shim (`EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX`)**: Emits Wormhole messages efficiently, without creating new message accounts for each emission, reducing rent costs.

Reviewer 2: Maybe add a link to the [program](https://explorer.solana.com/address/EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX)?

## Key Solana Concepts

To understand how shims work, it helps to know a few Solana basics:

- **PDA (Program Derived Address)**: 
    - Deterministic, program-owned accounts created without private keys.
    - In the legacy model, each message uses a unique PDA (derived from emitter and sequence number) to store data.
    - Shim difference: Emission shim skips PDA creation and instead puts message data in transaction logs, so nothing is left behind on-chain.
- **CPI (Cross-Program Invocation)**: Solana’s version of one smart contract calling another. Used by shims to invoke logic in the Core Bridge or other programs. 
- **[Anchor CPI Event](https://www.anchor-lang.com/docs/basics/cpi){target=\_blank}**: 
    - A structured log emitted during a CPI call, observable in transaction logs, used by the emission shim to report sequence number, timestamp, and payload.
    - The emission shim emits these events so Guardians can observe messages directly from transaction logs (rather than accounts).

## Guardian Observation Methods

> |                      | Legacy Model.                  | Shim Model                            |

Reviewer 2: ```suggestion
|                      | Legacy Model                   | Shim Model                            |
```

|----------------------|--------------------------------|---------------------------------------|
| Message Storage      | On-chain message account (PDA) | Transaction logs (CPI event)          |
| Data Permanence      | On-chain forever               | In logs (until RPC history is pruned) |
| Guardian Observation | Reads account data             | Reads transaction logs                |
| Cost                 | High (rent + compute)          | Low (compute only, no rent)           |
| Sequence Handling    | Account-based                  | Event-based                           |
| Closing Accounts     | Not possible                   | Not needed                            |

With shims, the message’s existence depends on the transaction log, so cost drops, but indefinite on-chain visibility is no longer guaranteed.

## Transaction Costs

Solana charges for two main resources when processing transactions: compute units (for execution) and rent (for storing data on-chain). Understanding how each contributes to overall cost is key to seeing why shims are so much cheaper.

> - **Compute Units (CU)**: Solana measures CPU resource usage per transaction as “compute units.” Each transaction has a CU limit (usually ~200,000, can be increased for a fee).

Reviewer 2: ```suggestion
- **Compute Units (CU)**: Solana measures CPU resource usage per transaction as “compute units”. Each transaction has a CU limit (usually ~200,000, can be increased for a fee).
```

---
title: Solana Shims
description: Understand how Wormhole uses shim programs on Solana to optimize message emission and VAA verification without modifying the Core Bridge.
categories: Basics
---

# Solana Shims

Wormhole shims on Solana are lightweight programs that enable cheaper and more flexible message emission and verification while preserving Guardian observation guarantees. They are designed for integrators who want to reduce Solana rent costs without sacrificing core protocol security or Guardian compatibility.

This page explains what shims are, why they were created, how they work, and what this means for integrators.

## The Core Bridge Account Problem

> When you emit a message on Solana using the legacy [Wormhole Core Bridge](/docs/protocol/infrastructure/core-contracts/){target=\_blank}, it creates a new on-chain account—a Program Derived Address (PDA)—for every message. Each of these accounts must hold enough SOL to be rent-exempt, locking up lamports that cannot be reclaimed since Core Bridge does not allow these accounts to be closed. Over time, this results in two big problems:

Reviewer 4: ```
account—a Program Derived Address (PDA)—for
``` 
is missing spaces around the dashes. should be
```
account — a Program Derived Address (PDA) — for
```

- **Permanent On-Chain State**: Every message leaves behind a permanent account, increasing long-term storage needs on Solana.
- **Lost Lamports to Rent**: Integrators lose SOL for every message, as the lamports needed for rent exemption remain locked in the message accounts indefinitely.

Solana’s rent-exemption model is designed to ensure account persistence, but this is not a limitation of the protocol itself. The real constraint is the legacy `post_message` function, which always creates a new, non-reclaimable account per emission. Even after a message is consumed, these accounts cannot be closed or reused, resulting in unrecoverable rent costs.

> Although the `post_message_unreliable` function allows for account reuse, it has strict limitations: overwritten messages are lost forever, and reused accounts can cause sequence conflicts or duplicate VAAs. Guardians expect every message to have a unique emitter and sequence, so reusing accounts risks breaking core protocol guarantees.

Reviewer 4: ```
reused accounts can cause sequence conflicts or duplicate VAAs
```

Where was this information gathered from? I'm not sure that the account reuse would cause sequence conflicts and I'm unsure what a "duplicate" VAA is. The main issue that I'm aware of is that the `unreliable` messages are not re-observable, so any miss from the guardians will be permanent.

Verification adds even more cost: the `post_vaa` instruction creates even more temporary accounts for signatures and VAA data, further increasing rent costs and on-chain state. These accounts aren’t automatically cleaned up, so the cost and on-chain state only grow with usage.

> This design does ensure reliability, messages and verification data are always available on-chain for Guardians to observe. However, it comes at a cost in both storage and lost SOL. To address these issues, Wormhole introduces Solana shims, which fundamentally change the cost model for emission and verification.

Reviewer 4: Guardians don't observe verification data. can probably just say `message data is always available...`

> ## What Are the Solana Shim Contracts?

Reviewer 2: ```suggestion
## What Are the Solana Shim Programs?
```

Shouldn't they be called programs since we are talking about Solana?

To address the limitations of the Core Bridge, Wormhole deploys two specialized Solana programs called shims:

- **[Post Message Shim (`EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX`)](https://explorer.solana.com/address/EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX){target=\_blank}**: Emits Wormhole messages efficiently, without creating new message accounts for each emission, reducing rent costs.
- **[Verify VAA Shim (`EFaNWErqAtVWufdNb7yofSHHfWFos843DFpu4JBw24at`)](https://explorer.solana.com/address/EFaNWErqAtVWufdNb7yofSHHfWFos843DFpu4JBw24at){target=\_blank}**: Verifies VAAs on-chain without leaving permanent accounts.

> Both shims act as lightweight wrappers around the existing Core Bridge. No upgrade to the main contract is required; Guardian infrastructure continues to work exactly as before.

Reviewer 4: If we're getting technical, [Guardian changes were required](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/programs/post-message/README.md#guardian-changes) for the post message shim, but it is correct to state the no upgrade to the Core Bridge program was required.

### “Shim” Emission and Verification

Wormhole shims on Solana refer to two different approaches depending on whether you are emitting messages or verifying VAAs:

> - Emission Shim: The emission shim is an actual Solana program that you must deploy. It wraps the Core Bridge’s `post_message_unreliable` instruction and emits message data as a log event. Integrators send messages through this program to avoid rent costs and state bloat. You must deploy and interact with the emission shim as a separate contract.

Reviewer 4: ```
The emission shim is an actual Solana program that you must deploy.
```

This is not accurate. There is only one canonical Post Message shim that the Guardians are monitoring, `EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX`, and it has already been deployed. Integrators may require changes to their existing programs should they chose to use it, but they do not need to deploy this program.

> - Verification “Shim”: The verification shim is not a new program or contract. Instead, it refers to a rent-efficient usage pattern of the existing Core Bridge instructions (`verify_signatures` and `post_vaa`). In this flow, you create the minimum required temporary accounts for VAA verification, perform your checks, then immediately close those accounts to reclaim rent. All logic is handled by the standard Core Bridge contract—no additional contract deployment is needed.

Reviewer 4: ```
The verification shim is not a new program or contract.
```

This is not accurate. As linked above, the Verification shim is a program at `EFaNWErqAtVWufdNb7yofSHHfWFos843DFpu4JBw24at`. That is the program which you can interact with to verify VAA signatures in place of the Core Bridge's `verify_signatures` and `post_vaa`. 

The logic is *not* handled by the standard Core Bridge, it is contained in the `verify_hash` instruction of the shim, located [here](https://github.com/wormhole-foundation/wormhole/blob/4656bd4a72cb99f4e94a771a802856c9451af844/svm/wormhole-core-shims/programs/verify-vaa/src/lib.rs#L195C4-L195C23). 

While it is correct to state that no additional contract deployment is needed. Existing integrators would need to make changes to adopt the new approach.

> | Purpose            | Is it a new program? | Deployment Required? | How it works                                |

Reviewer 4: Both of these are new programs, neither of them should be deployed by integrators.

|--------------------|----------------------|----------------------|---------------------------------------------|

> | Emission Shim      | Yes                  | Yes                  | Deploy the shim, call its instruction       |

Reviewer 4: How -> call `post_message` on the Post Message shim. Here's the [IDL](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/idls/wormhole_post_message_shim.json)

> | Verification Shim  | No                   | No                   | Use standard Core Bridge; close temp accounts after verification |

Reviewer 4: How -> first call `post_signatures` on the Verification Shim. Then, from within your program, call `verify_hash` (see [README](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/programs/verify-vaa/README.md#verify-hash-technical-details) for computing the v1 VAA digest). 

In the same transaction, append an instruction to close the signature account, reclaiming the lamports with `close_signatures`. 

Here's the [IDL](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/idls/wormhole_verify_vaa_shim.json)

## Key Solana Concepts

To understand how shims work, it helps to know a few Solana basics:

- **PDA (Program Derived Address)**: 
    - Deterministic, program-owned accounts created without private keys.

>     - In the legacy model, each message uses a unique PDA (derived from emitter and sequence number) to store data.

Reviewer 4: This is not accurate. The Core Bridge does not use a PDA for message accounts, it requires a keypair to be generated for every new message account.

>     - Shim difference: Emission shim skips PDA creation and instead puts message data in transaction logs, so nothing is left behind on-chain.

Reviewer 4: This is not exactly correct. It avoids requiring a new keypair by using [a PDA generated by the Post Message shim](https://github.com/wormhole-foundation/wormhole/blob/4656bd4a72cb99f4e94a771a802856c9451af844/svm/wormhole-core-shims/programs/post-message/src/lib.rs#L62).

- **CPI (Cross-Program Invocation)**: Solana’s version of one smart contract calling another. Used by shims to invoke logic in the Core Bridge or other programs. 
- **[Anchor CPI Event](https://www.anchor-lang.com/docs/basics/cpi){target=\_blank}**: 
    - A structured log emitted during a CPI call, observable in transaction logs, used by the emission shim to report sequence number, timestamp, and payload.
    - The emission shim emits these events so Guardians can observe messages directly from transaction logs (rather than accounts).

## Guardian Observation Methods

|                      | Legacy Model                   | Shim Model                            |
|----------------------|--------------------------------|---------------------------------------|
| Message Storage      | On-chain message account (PDA) | Transaction logs (CPI event)          |
| Data Permanence      | On-chain forever               | In logs (until RPC history is pruned) |
| Guardian Observation | Reads account data             | Reads transaction logs                |
| Cost                 | High (rent + compute)          | Low (compute only, no rent)           |

> | Sequence Handling    | Account-based                  | Event-based                           |

Reviewer 4: The sequence handling is unchanged and still uses the same account. The integrator can freely switch between both methods while maintaining their sequence number.

## Transaction Costs

Solana charges for two main resources when processing transactions: compute units (for execution) and rent (for storing data on-chain). Understanding how each contributes to overall cost is key to seeing why shims are so much cheaper.

- **Compute Units (CU)**: Solana measures CPU resource usage per transaction as “compute units”. Each transaction has a CU limit (usually ~200,000, can be increased for a fee).
- **Rent**: One-time cost in SOL to keep an account on-chain. Most of the Core Bridge’s cost comes from rent, not CUs.

!!!note "Why is the shim cheaper?"
    Even though the shim uses slightly more compute (extra logic for logging), it avoids account creation entirely. Since rent is the most significant cost, the total emission cost drops.

## Safety, Tradeoffs & Limitations

Shims keep all security guarantees of the Core Bridge, except that messages and verification data are not permanently stored on-chain. If a Guardian misses the transaction log (for example, if their node history is too short), re-observation is only possible as long as the transaction history is available.

> While the Solana shims dramatically reduce costs and prevent long-term account bloat, they do so at the expense of message permanence. With shims, messages and verification data are no longer stored in permanent on-chain accounts; instead, they exist only in transaction logs for a limited period, meaning Guardians must observe these logs promptly or risk missing the message once node history is pruned. This means Guardians must observe these logs promptly, or the opportunity to process the message may be lost once the node’s transaction history is pruned. As a result, shims are ideal for integrators who prioritize cost savings and efficiency over indefinite on-chain availability. However, for use cases that require permanent on-chain message history and auditability, the legacy Core Bridge approach remains the safer choice, despite its higher rent costs and storage requirements.

Reviewer 4: For what it's worth, none of the other Wormhole implementations permanently store messages on chain and the logs are available via the explorers (should that be sufficient for "auditability")
